### PR TITLE
Mk 1938 additional unit tests for forger 

### PR DESCRIPTION
**Pull request by michaelkernaghan**
CLOSE #19381

extend unit test coverage for forger package

### CODE REVIEW

File: packages/taquito-local-forging/test/taquito-local-forging.spec.ts [hunk 1]
```diff
@@ -19,153 +20,255 @@ describe('Forge and parse operations default protocol', () => {
...
+      const decodeCodec = myGetCodec.decoder('0572cbea904d67468808c8eb50a9450c9721db309128012543902d0ac358a62a')
+      expect(decodeCodec).toEqual("0572cbea904d67468808c8eb50a9450c9721db309128012543902d0ac358a62a")
+      done();
+    });
+
+    test(`Verify Arrow Function for CODEC.OP_ACTIVATE_ACCOUNT`, async (done) => {
+      const codec = CODEC.OP_ACTIVATE_ACCOUNT
+      const myGetCodec = getCodec(codec, ProtocolsHash.PtKathman)
+      const gotCodec = myGetCodec.decoder('0572cbea904d67468808c8eb50a9450c9721db309128012543902d0ac358a62a')
+      expect(gotCodec).toEqual({ "pkh": "tz1L8qbHKzqFmLiZqxaJdMEWrCwmmrSw2wj4", "secret": "9128012543902d0ac358a62a" })
```
Comment (roxaneletourneau) [inline packages/taquito-local-forging/test/taquito-local-forging.spec.ts:237]: Secret seems too short, are we missing a validation there? Do we need a follow-up ticket?

File: packages/taquito-local-forging/test/taquito-local-forging.spec.ts [hunk 2]
```diff
@@ -19,153 +20,255 @@ describe('Forge and parse operations default protocol', () => {
...
+
+      const consumer = Uint8ArrayConsumer.fromHexString(hexToParse)
+      expect(decoders[codec](consumer)).toStrictEqual({ "pkh": "tz1L8qbHKzqFmLiZqxaJdMEWrCwmmrSw2wj4", "secret": "9128012543902d0ac358a62ae28f75bb8f1c7c42" })
+      expect((schemaDecoder(decoders)(ActivationSchema)(consumer))).toBeDefined
+      expect((schemaDecoder(decoders)(ActivationSchema)(consumer))).toStrictEqual({ "pkh": "8wkhLGYvWfaqQXYRo5yCvSe4mVr314j", "secret": "" })
```
Comment (roxaneletourneau) [inline packages/taquito-local-forging/test/taquito-local-forging.spec.ts:242]: Why is secret empty, are we missing a validation?

File: packages/taquito-local-forging/test/taquito-local-forging.spec.ts [hunk 3]
```diff
@@ -19,153 +20,255 @@ describe('Forge and parse operations default protocol', () => {
...
+      done();
+    });
+
+    test(`test getCodec to verify codec Manager sent to decoders is defined`, async (done) => {
+      const myGetCodec = getCodec(CODEC.MANAGER, ProtocolsHash.PtKathman).decoder('7c842c15c8b0c8fd228e6cb5302a50201f41642dd36b699003fb3c857920bc9d')
+      expect(myGetCodec).toEqual({ "branch": "BLf7wKNryZRXibjzM4TjNBSMgNN4qJVhxRRuxo3uu3SegsFqkUd", "contents": [] })
+      done();
+    });
+
+    test(`test getCodec to verify codec secret sent to decoders is defined`, async (done) => {
```
Comment (roxaneletourneau) [inline packages/taquito-local-forging/test/taquito-local-forging.spec.ts:252]: Kind if duplicate of `Verify Arrow Functions for CODEC.SECRET is toHexString(val.consume(20)`

**Review by roxaneletourneau — CHANGES_REQUESTED** [review]

File: packages/taquito-local-forging/test/codec.spec.ts [hunk 4]
```diff
@@ -40,3 +42,54 @@ describe('Entrypoint', () => {
...
 });
+
+test(`Verify pkhEncoder`, async (done) => {
+  const tz1 = pkhEncoder('tz1e42w8ZaGAbM3gucbBy8iRypdbnqUj7oWY')
+  expect(tz1).toEqual('00c9fc72e8491bd2973e196f04ec6918ad5bcee22d')
+  const tz2 = pkhEncoder('tz2Ch1abG7FNiibmV26Uzgdsnfni9XGrk5wD')
+  expect(tz2).toEqual('012ffebbf1560632ca767bc960ccdb84669d284c2c')
+  const tz3 = pkhEncoder('tz3WXYtyDUNL91qfiCJtVUX746QpNv5i5ve5')
+  expect(tz3).toEqual('026fde46af0356a0476dae4e4600172dc9309b3aa4')
+  try {
```
[the remark this answers is not in the document]
Comment (michaelkernaghan) [reply packages/taquito-local-forging/test/codec.spec.ts:53]: changed to suit the throw formatting

File: packages/taquito-local-forging/test/codec.spec.ts [hunk 5]
```diff
@@ -40,3 +42,54 @@ describe('Entrypoint', () => {
...
+
+test(`Verify publicKeyEncoder`, async (done) => {
+  const edpk = publicKeyEncoder('edpkuLxx9PQD8fZ45eUzrK3BhfDZJHhBuK4Zi49DcEGANwd2rpX82t')
+  expect(edpk).toEqual('005c8244b8de7d57795962c1bfc855d0813f8c61eddf3795f804ccdea3e4c82ae9')
+  const sppk = publicKeyEncoder('sppk7aVdgAmezMCRTcHciVkVZoGNnhSdKEYcn5pYaqt4PvLjgFbLRxo')
+  expect(sppk).toEqual('01029caa79bfddd2bc198ff8b408e6fde1a10accb59a8148b1958f175d83b957b579')
+  const p2pk = publicKeyEncoder('p2pk67c5b5THCj5fyksX1C13etdUpLR9BDYvJUuJNrxeGqCgbY3NFpV')
+  expect(p2pk).toEqual('02038f711262480454088fe0d31254211e04cb785affcc9280ce719e3f7e763f974d')
+  try {
+    publicKeyEncoder('p4zzk67c5b5THCj5fyksX1C13etdUpLR9BDYvJUuJNrxeGqCgbY3NFpV')
```
[the remark this answers is not in the document]
Comment (michaelkernaghan) [reply packages/taquito-local-forging/test/codec.spec.ts:70]: changed to suit the throw formatting

File: packages/taquito-local-forging/test/codec.spec.ts [hunk 6]
```diff
@@ -40,3 +42,54 @@ describe('Entrypoint', () => {
...
+
+  const edpk = publicKeyDecoder(Uint8ArrayConsumer.fromHexString('005c8244b8de7d57795962c1bfc855d0813f8c61eddf3795f804ccdea3e4c82ae9'))
+  expect(edpk).toEqual('edpkuLxx9PQD8fZ45eUzrK3BhfDZJHhBuK4Zi49DcEGANwd2rpX82t')
+
+  const sppk = publicKeyDecoder(Uint8ArrayConsumer.fromHexString('01029caa79bfddd2bc198ff8b408e6fde1a10accb59a8148b1958f175d83b957b579'))
+  expect(sppk).toEqual('sppk7aVdgAmezMCRTcHciVkVZoGNnhSdKEYcn5pYaqt4PvLjgFbLRxo')
+ 
+  const p2pk = publicKeyDecoder(Uint8ArrayConsumer.fromHexString('02038f711262480454088fe0d31254211e04cb785affcc9280ce719e3f7e763f974d'))
+  expect(p2pk).toEqual('p2pk67c5b5THCj5fyksX1C13etdUpLR9BDYvJUuJNrxeGqCgbY3NFpV')
+  try {
```
[the remark this answers is not in the document]
Comment (michaelkernaghan) [reply packages/taquito-local-forging/test/codec.spec.ts:88]: changed to use the throw format

File: packages/taquito-local-forging/test/forger-errors.spec.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import { ballotDecoder, ballotEncoder, entrypointDecoder } from '../src/codec';
+import { DecodeBallotValueError, InvalidBallotValueError, InvalidHexStringError, OversizedEntryPointError} from '../src/error';
+import { bytesEncoder } from '../src/michelson/codec';
+import { Uint8ArrayConsumer } from '../src/uint8array-consumer';
+
+
+describe('Tests for errors thrown when forging fails', () => {
```
[the remark this answers is not in the document]
Comment (michaelkernaghan) [reply packages/taquito-local-forging/test/forger-errors.spec.ts:7]: Moved the tests to codec.spec.ts  to be with the others and converted try catch to Throw format.

File: packages/taquito-local-forging/test/forger-errors.spec.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    test('Verify that ballotDecoder functions correctly and returns DecodeBallotValueError on unknown case ', () => {
+        expect(ballotDecoder(Uint8ArrayConsumer.fromHexString('005c8244b8de7d57795962c1bfc855d0813f8c61eddf3795f804ccdea3e4c82ae9'))).toEqual('yay')
+        expect(ballotDecoder(Uint8ArrayConsumer.fromHexString('01029caa79bfddd2bc198ff8b408e6fde1a10accb59a8148b1958f175d83b957b579'))).toEqual('nay')
+        expect(ballotDecoder(Uint8ArrayConsumer.fromHexString('02038f711262480454088fe0d31254211e04cb785affcc9280ce719e3f7e763f974d'))).toEqual('pass')
+        try {
+            ballotDecoder(Uint8ArrayConsumer.fromHexString('03038f711262480454088fe0d31254211e04cb785affcc9280ce719e3f7e763f974d'))
+        } catch (e: any) {
+            expect(e).toBeInstanceOf(DecodeBallotValueError);
+            expect(e.name).toEqual('DecodeBallotValueError');
+            expect(e.message).toEqual("Failed to decode ballot value 3");
```
[the remark this answers is not in the document]
Comment (michaelkernaghan) [reply packages/taquito-local-forging/test/forger-errors.spec.ts:31]: The 3 is from the leading digits I put on the string that I want to provoke a fail.  romHexString('03038f7
the function is looking for 00,01,02, so 03 gives the error

File: packages/taquito-local-forging/test/taquito-local-forging.spec.ts [hunk 9]
```diff
@@ -19,154 +20,255 @@ describe('Forge and parse operations default protocol', () => {
...
+      expect(consumer).toEqual('0572cbea904d67468808c8eb50a9450c9721db30')
+      done();
+    });
 
-  describe('Verify the ProtoInferiorTo function in protocols.ts', () => {
+    test(`Verify getCodec for CODEC.RAW`, async (done) => {
+      const codec = CODEC.RAW
+      const myGetCodec = getCodec(codec, ProtocolsHash.PtKathman)
+      const consumer = myGetCodec.decoder(hexToParse)
+      expect(consumer).toBeDefined
```
[the remark this answers is not in the document]
Comment (michaelkernaghan) [reply packages/taquito-local-forging/test/taquito-local-forging.spec.ts:189]: changed to toBeDefined()

File: packages/taquito-local-forging/test/taquito-local-forging.spec.ts [hunk 10]
```diff
@@ -19,154 +20,255 @@ describe('Forge and parse operations default protocol', () => {
...
+        expect(e.name).toEqual('UnsupportedOperationError');
+        expect(e.message).toEqual(`The operation '76' is unsupported`);
+      }
+    });
+
+    test(`Verify getCodec for CODEC.SECRET`, async (done) => {
+      const codec = CODEC.SECRET
+      const myGetCodec = getCodec(codec, ProtocolsHash.PtKathman)
+      const consumer = myGetCodec.decoder(hexToParse)
+      expect(consumer).toBeDefined
```
[the remark this answers is not in the document]
Comment (michaelkernaghan) [reply packages/taquito-local-forging/test/taquito-local-forging.spec.ts:180]: changed to toBeDefined()

File: packages/taquito-local-forging/test/taquito-local-forging.spec.ts [hunk 11]
```diff
@@ -19,154 +20,255 @@ describe('Forge and parse operations default protocol', () => {
...
+      done();
+    });
 
-    const a = ProtocolsHash.Pt24m4xi
-    const b = ProtocolsHash.PtKathman
+    test(`Verify getCodec for CODEC.OP_DELEGATION`, async (done) => {
+      const codec = CODEC.OP_DELEGATION
+      const myGetCodec = getCodec(codec, ProtocolsHash.PtKathman)
+      const consumer = myGetCodec.decoder(hexToParse)
+      expect(consumer).toEqual({ "counter": "161756491", "fee": "114", "gas_limit": "103", "storage_limit": "70" })
```
[the remark this answers is not in the document]
Comment (michaelkernaghan) [reply packages/taquito-local-forging/test/taquito-local-forging.spec.ts:198]: I tried to verify and add coverage for the getCodec(codec, ProtocolsHash.PtKathman) function. It gets the codec from the constants.ts file.  I am unsure how to use the schema you give in the test.

File: packages/taquito-local-forging/test/taquito-local-forging.spec.ts (hunk 1 above)
Comment (michaelkernaghan) [reply packages/taquito-local-forging/test/taquito-local-forging.spec.ts:238]: The hex string i was encoding was not complete. It was changed to use a proper hex string. Now the secret is the correct length.

File: packages/taquito-local-forging/test/taquito-local-forging.spec.ts (hunk 2 above)
Comment (michaelkernaghan) [reply packages/taquito-local-forging/test/taquito-local-forging.spec.ts:243]: The pkh calculated is also wrong. I think I am in the test encoding something twice, so the second encoding is gibberish. Dropped the extra line with the additional encoding check as it is meaningless.

Comment (michaelkernaghan) [reply packages/taquito-local-forging/test/taquito-local-forging.spec.ts:243]: I put that test in there to try to kill a mutant. It did not kill any mutants, so there is no loss in removing the line.

File: packages/taquito-local-forging/test/taquito-local-forging.spec.ts (hunk 3 above)
Comment (michaelkernaghan) [reply packages/taquito-local-forging/test/taquito-local-forging.spec.ts:253]: True. Deleted the dupe.

File: packages/taquito-contracts-library/test/taquito-contracts-library.spec.ts [hunk 12]
```diff
@@ -83,39 +83,35 @@ describe('ContractsLibrary tests', () => {
...
         },
-      });
-    } catch (e: any) {
-      expect(e).toBeInstanceOf(InvalidAddressError);
-      expect(e.message).toEqual(`Address is invalid: ${contractAddress}`);
-      expect(e).toBeInstanceOf(Error);
-    }
+      })
+    ).toThrow(InvalidAddressError);
+    expect.objectContaining({
```
Comment (roxaneletourneau) [inline packages/taquito-contracts-library/test/taquito-contracts-library.spec.ts:94]: Something is missing here. `expect.objectContaining` cannot be called just like that. We can change the expected string, and the test will still pass.

**Review by roxaneletourneau — CHANGES_REQUESTED** [review]

File: packages/taquito-contracts-library/test/taquito-contracts-library.spec.ts (hunk 12 above)
Comment (michaelkernaghan) [reply packages/taquito-contracts-library/test/taquito-contracts-library.spec.ts:94]: i have fixed the examples where this mistake was made

File: packages/taquito-local-forging/test/codec.spec.ts [hunk 13]
```diff
@@ -39,4 +48,144 @@ describe('Entrypoint', () => {
...
+
+  test(`Verify pkhEncoder`, async (done) => {
+    const tz1 = pkhEncoder('tz1e42w8ZaGAbM3gucbBy8iRypdbnqUj7oWY');
+    expect(tz1).toEqual('00c9fc72e8491bd2973e196f04ec6918ad5bcee22d');
+    const tz2 = pkhEncoder('tz2Ch1abG7FNiibmV26Uzgdsnfni9XGrk5wD');
+    expect(tz2).toEqual('012ffebbf1560632ca767bc960ccdb84669d284c2c');
+    const tz3 = pkhEncoder('tz3WXYtyDUNL91qfiCJtVUX746QpNv5i5ve5');
+    expect(tz3).toEqual('026fde46af0356a0476dae4e4600172dc9309b3aa4');
+    expect(() => pkhEncoder('tz4WXYtyDUNL91qfiCJtVUX746QpNv5i5ve5')).toThrow(InvalidKeyHashError);
+    expect.objectContaining({
```
Comment (roxaneletourneau) [inline packages/taquito-local-forging/test/codec.spec.ts:60]: Missing something, there are other occurrences in the file that needs to be fixed

**Review by roxaneletourneau — CHANGES_REQUESTED** [review]

File: packages/taquito-local-forging/test/taquito-local-forging.spec.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (michaelkernaghan) [reply packages/taquito-local-forging/test/taquito-local-forging.spec.ts]: For these rets, it won't matter as the hex string is arbitrary for these test, which are mostly trying to invoke error conditions for test coverage.

File: packages/taquito-local-forging/test/codec.spec.ts (hunk 13 above)
Comment (michaelkernaghan) [reply packages/taquito-local-forging/test/codec.spec.ts]: fixed all of the incorrecty written tests of the exception message thrown

**Review by roxaneletourneau — CHANGES_REQUESTED** [review]

File: packages/taquito-local-forging/test/forger-errors.spec.ts [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+  DecodeBallotValueError,
+  InvalidBallotValueError,
+  InvalidHexStringError,
+  OversizedEntryPointError,
+} from '../src/error';
+import { bytesEncoder } from '../src/michelson/codec';
+import { Uint8ArrayConsumer } from '../src/uint8array-consumer';
+
+describe('Tests for errors thrown when forging fails', () => {
+  test('Verify that ballotEncoder functions correctly and returns InvalidBallotValueError on unknown case ', () => {
```
[the remark this answers is not in the document]
Comment (michaelkernaghan) [reply packages/taquito-local-forging/test/forger-errors.spec.ts:12]: Yes, they are duplicated, deleted the duped file.

File: packages/taquito-local-forging/test/codec.spec.ts [hunk 15]
```diff
@@ -39,4 +48,184 @@ describe('Entrypoint', () => {
...
+
+    expect(() =>
+      publicKeyDecoder(
+        Uint8ArrayConsumer.fromHexString(
+          '035c8244b8de7d57795962c1bfc855d0813f8c61eddf3795f804ccdea3e4c82ae9'
+        )
+      )
+    ).toThrow(
+      expect.objectContaining({
+        message: expect.stringContaining("The public key '[object Object]' is invalid."),
```
[the remark this answers is not in the document]
Comment (michaelkernaghan) [reply packages/taquito-local-forging/test/codec.spec.ts:130]: ok

**Review by roxaneletourneau — APPROVED** [review]

**Pull request merged**